rust/kernel/str.rs

```diff
@@ -0,0 +1,243 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! String representations.
+
+use core::ops::{self, Deref, Index};
+
+use crate::bindings;
+use crate::c_types;
+
+/// Byte string without UTF-8 validity guarantee.
+///
+/// `BStr` is simply an alias to `[u8]`, but has a more evident semantical meaning.
+pub type BStr = [u8];
+
+/// Creates a new [`BStr`] from a string literal.
+///
+/// `b_str!` converts the supplied string literal to byte string, so non-ASCII
+/// characters can be included.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// const MY_BSTR: &'static BStr = b_str!("My awesome BStr!");
+/// ```
+#[macro_export]
+macro_rules! b_str {
+    ($str:literal) => {{
+        const S: &'static str = $str;
+        const C: &'static $crate::str::BStr = S.as_bytes();
+        C
+    }};
+}
+
+/// Possible errors when using conversion functions in [`CStr`].
+#[derive(Debug, Clone, Copy)]
+pub enum CStrConvertError {
+    /// Supplied bytes contain an interior `NUL`.
+    InteriorNul,
+
+    /// Supplied bytes are not terminated by `NUL`.
+    NotNulTerminated,
+}
+
+impl From<CStrConvertError> for crate::Error {
+    #[inline]
+    fn from(_: CStrConvertError) -> crate::Error {
+        crate::Error::EINVAL
+    }
+}
+
+/// A string that is guaranteed to have exactly one `NUL` byte, which is at the
+/// end.
+///
+/// Used for interoperability with kernel APIs that take C strings.
+#[repr(transparent)]
+pub struct CStr([u8]);
+
+impl CStr {
+    /// Returns the length of this string excluding `NUL`.
+    #[inline]
+    pub const fn len(&self) -> usize {
+        self.0.len() - 1
+    }
+
+    /// Returns the length of this string with `NUL`.
+    #[inline]
+    pub const fn len_with_nul(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if the string only includes `NUL`.
+    #[inline]
+    pub const fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Wraps a raw C string pointer.
+    ///
+    /// # Safety
+    ///
+    /// `ptr` must be a valid pointer to a `NUL`-terminated C string, and it must
+    /// last at least `'a`. When `CStr` is alive, the memory pointed by `ptr`
+    /// must not be mutated.
+    #[inline]
+    pub unsafe fn from_char_ptr<'a>(ptr: *const c_types::c_char) -> &'a Self {
+        let len = bindings::strlen(ptr) + 1;
+        Self::from_bytes_with_nul_unchecked(core::slice::from_raw_parts(ptr as _, len as _))
+    }
+
+    /// Creates a [`CStr`] from a `[u8]`.
+    ///
+    /// The provided slice must be `NUL`-terminated, does not contain any
+    /// interior `NUL` bytes.
+    pub const fn from_bytes_with_nul(bytes: &[u8]) -> Result<&Self, CStrConvertError> {
```

Comment:
```suggestion
    pub const fn try_from_bytes_with_nul(bytes: &[u8]) -> Result<&Self, CStrConvertError> {
```

Reply:
Rust std does not usually use `try_` prefix for constructors that may fail.
* https://doc.rust-lang.org/std/str/fn.from_utf8.html
* https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_bytes_with_nul
* https://doc.rust-lang.org/std/num/struct.NonZeroUsize.html#method.new

It seems that in Rust std, `try_` is only added if a method without `try_` prefix exists already, e.g. `from` vs `try_from`, `fold` vs `try_fold`.

Reply:
The `from_bytes_with_nul_unwrap` is the counterpart to this function. If this function starts with `try_` the other doesn't need `_unwrap` was my reasoning.

Reply:
Well, creating a `CStr` from `[u8]` is inherently a fallible operation, and `from_bytes_with_nul` should always be used. `from_bytes_with_nul_unwrap` is only there because `const_panic` is not yet stable and I don't want all users of `c_str!` to have `#![feature(const_panic)]`.

Once `const_panic` is stablised, `from_bytes_with_nul_unwrap` should be removed and have the check be baked into `c_str!` instead. `from_bytes_with_nul_unwrap` is only a temporary solution (and thus is `#[doc(hidden)]`).